elyra/pipeline/component.py

```diff
@@ -252,13 +254,46 @@ def resource_type(self):
         return self.location_type
 
     @abstractmethod
-    def read_component_definition(self, location: str) -> Optional[str]:
+    def read_component_definition(self,
+                                  location: str,
+                                  location_to_def: Dict[str, str]) -> Dict[str, str]:
         """
         Read an absolute location to get the contents of a component
         specification file
+
+        :param location: an absolute path to the specification file to read
+        :param location_to_def: a mapping of component locations to definitions
+
+        :returns: the given 'location_to_def' object, optionally including a new
+                  key-value pair if the given component location is successfully read
         """
         raise NotImplementedError()
 
+    def read_component_definitions(self, locations: List[str]) -> Dict[str, str]:
+        """
+        This function starts a thread for each of the given locations in order to read
+        component definitions in parallel.
+
+        The 'location_to_def' variable is a mapping of a component location to its content.
+        As a mutable object, this dictionary provides a means to retrieve a return value for
+        each thread. If a thread is able to successfully read the content of the given
+        component file location, a location-to-content mapping is added to 'location_to_def'.
+        """
+        location_to_def = {}
+
+        # Start a reader thread for each location
+        threads = []
+        for location in self.get_absolute_locations(locations):
+            t = Thread(target=self.read_component_definition, args=(location, location_to_def))
+            threads.append(t)
+            t.start()
+
+        # Wait for all threads to finish
+        for t in threads:
+            t.join()
```

Comment:
I think it would be a good idea to go ahead and limit the number of threads to a configurable amount (thinking a default of 3).  Since `ComponentReader` is a `LoggingConfigurable` this could be easily expressed via traitlets, e.g., `max_reader_threads`.  (Here's an [example of a configured trait from EG](https://github.com/jupyter/enterprise_gateway/blob/d6d6ff59fb33ddb9916f1e396f5e335d5a3dc6b8/enterprise_gateway/mixins.py#L286-L295).)

To implement, we'd toss the locations into a queue and use a semaphore (initialized to `min(queueLen, max_threads)`) to gate the threads and let the threads drain the queue.

Of course, this won't have the overall performance increase as thread-per-location but it will avoid any pathological cases that arise with larger registries.